feat: support prepare statement

NGA-TRAN opened the pull request:
# Which issue does this PR close?
<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/arrow-datafusion/issues/4426

Re: https://github.com/apache/arrow-datafusion/issues/4539

- [x] This PR  generates a LogicalPlan for `PREPARE SELECT ...`
- [ ] `EXECUTE statement` will be done in a follow-on PR

# Which prepare statements supported in this PR:

- [x] Params in WHERE clause including both `column > $1` and `column in [$1, $2, ...]`
- [x] Params in HAVING clause including both `AGG(column) > $1` and `AGG(column) in [$1, $2, ...]` 
- [x] Params in VALUES table in  SELECT. Eg: `SELECT * FROM  (VALUES(1, $1), (2, $2)) AS t (num, letter);`

Those are all I can think of for `SELECT`. Let me know if you think of other ones and I will add them in the tests. This PR does NOT support params in UPDATE and INSERT

# What changes are included in this PR?

- Add LogicalPlan::Prepare
- Support Expr::Placeholder as it is a part of Prepare stmt
- Generate LogicalPlan::Prepare for `PREPARE stmt`
- Generate corresponding protobuf messages for LogicalPlan::Prepare
- Map given data types of the PREPARE stmt to their corresponding Placeholders in the stmt

# Are these changes tested?

Yes

# Are there any user-facing changes?

Yes, they can use PREPARE stmt now

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

NGA-TRAN commented on datafusion/sql/src/planner.rs:
```diff
@@ -5992,6 +6011,20 @@ mod tests {
+
+        quick_test(sql, expected);
+    }
+
```
This is the only test for now. I will add more tests

alamb replied:
The other thing I don't see is code to figure out the placeholders and their types

NGA-TRAN commented:
@alamb, @thinkharderdev @metesynnada  I will add more tests but this PR has implemented the `PREPARE stmt` to create a logical plan from a sql string (see the test)

Can you have a quick look to see if it is what you expect while I add more tests?

I will implement the `EXECUTE stmt` either in the following PR (preferred) or in this PR if you think better.

NGA-TRAN commented on datafusion/proto/src/generated/datafusion.rs:
```diff
@@ -0,0 +1,1512 @@
+#[derive(Clone, PartialEq, ::prost::Message)]
```
I guess I do not need to commit this file, right?

alamb replied:
Correct -- it should not be part of the PR

alamb commented on datafusion/core/src/physical_plan/planner.rs:
```diff
@@ -344,6 +344,9 @@ fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
         Expr::QualifiedWildcard { .. } => Err(DataFusionError::Internal(
             "Create physical name does not support qualified wildcard".to_string(),
         )),
+        Expr::Placeholder(_) => Err(DataFusionError::Internal(
```
👍

alamb commented on datafusion/expr/src/expr_schema.rs:
```diff
@@ -198,7 +199,8 @@ impl ExprSchemable for Expr {
             | Expr::IsNotUnknown(_)
-            | Expr::Exists { .. } => Ok(false),
+            | Expr::Exists { .. }
+            | Expr::Placeholder(_) => Ok(false), // todo: Placeholder should return false?
```
I think since we don't know if the placeholder will be null, it should return true to be safe

alamb commented on datafusion/expr/src/logical_plan/builder.rs:
```diff
@@ -119,6 +121,7 @@ impl LogicalPlanBuilder {
     /// The column names are not specified by the SQL standard and different database systems do it differently,
     /// so it's usually better to override the default names with a table alias list.
     pub fn values(mut values: Vec<Vec<Expr>>) -> Result<Self> {
+        // todo: hanlde for Placeholder expr
```
yeah, in general the Values statement should probably be cleaned up / generalized. I can try to help with this next week if that is useful

alamb commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -108,6 +108,8 @@ pub enum LogicalPlan {
     /// Set a Variable
     SetVariable(SetVariable),
+    /// Prepare a statement
+    Prepare(Prepare),
```
Is the plan to add a `Execute` statement as a follow on PR?

NGA-TRAN replied:
I think so. It will be cleaner unless you say otherwise

alamb replied:
Follow on PR sounds good to me 👍

alamb commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1358,6 +1368,17 @@ pub struct CreateExternalTable {
     pub options: HashMap<String, String>,
 }
 
+/// Prepare a statement
```
```suggestion
/// Prepare a statement but do not execute it. Prepared statements
/// Can have 0 or more `Expr::Placeholder` expressions that are filled in
/// during execution. 
```

alamb commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1358,6 +1368,17 @@ pub struct CreateExternalTable {
+pub struct Prepare {
+    /// The name of the statement
+    pub name: String,
+    /// Data types of the parameters
```
```suggestion
    /// Data types of the parameters ([`Expr::PlaceHolder`])
```

alamb reviewed (commented):
This is looking exactly along the right track @NGA-TRAN  -- thank you !

alamb commented:
I plan to review this carefully tomorrow morning

metesynnada commented:
LGTM. As far as I can see, we do not store the plans, just use them immediately after we prepare. I hope I am not missing anything.

mingmwang commented:
I will take a look at this PR carefully tomorrow.
Generally, Prepared Statement and Plan Cache support is quite complex in RDBMS.

alamb commented on datafusion/expr/src/expr.rs:
```diff
@@ -244,6 +244,8 @@ pub enum Expr {
     /// GROUP BY expression list
     GroupingSet(GroupingSet),
+    /// A place holder for parameters in a prepared statement.
+    Placeholder(String),
```
@NGA-TRAN  -- I was thinking about how to get the data type information from the SQL planner to here

I was wondering what if you looked up the placeholder during planning and then it would be available for the rest of the system

Something like 

```suggestion
    Placeholder {
      /// The identifier
      id: String,
      /// The type the parameter will be filled in with
      data_type: DataType,
     },
```

NGA-TRAN replied:
I think this is a good idea. This way, we won't miss to get data type of any params

NGA-TRAN replied:
I am trying to do this and see if I can eliminate the work in get data type that will simplify this PR a lot

NGA-TRAN replied:
Done

NGA-TRAN commented on datafusion/expr/src/expr_schema.rs:
```diff
@@ -23,12 +23,20 @@ use crate::{aggregate_function, function, window_function};
     fn get_type<S: ExprSchema>(&self, schema: &S) -> Result<DataType>;
 
+    /// given a schema and param data types, return the type of the expr
+    fn get_type_with_params<S: ExprSchema>(
```
@alamb : Yes, at first I planned to add the parameter into `get_type` function but then I saw a massive change and I thought this solution would be preferable but I agree there is a trade-off that if we miss using it, it will lead to bugs.

I can make a massive change easily if reviewers agree to do so and I think it will solve all edge cases and won't lead to bugs because we pass it all the way down from statement where we know whether the parameter is (non-empty or empty vector of params of a PlannerContext)

Let me have a quick look to see if moving parameter into DFSchema is a good option. I am thinking it will lead to the same dilemma in the code somewhere

NGA-TRAN replied:
Since the data type is stored with the LogicalPlan::Prepare, we no longer need this

NGA-TRAN commented on datafusion/expr/src/logical_plan/builder.rs:
```diff
@@ -272,11 +279,31 @@ impl LogicalPlanBuilder {
+    }
+
+    ///  Apply a filter wit provided data types for params of prepared statement
+    pub fn filter_with_params(
```
We no longer need this after implementing @alamb's suggestion. Thanks @alamb

NGA-TRAN commented:
I am working on adding more tests which I think the last piece of this PR

NGA-TRAN commented:
@alamb : I have addressed all your comments. Can you re-review it?

Notes: If you run queries in DataFusion-cli, you will get error `Internal("Unsupported logical plan: Prepare")`. I think this means we are successfully create a LogicalPlan::Prepare (the purpose of this PR) but since we have not store it anywhere and have not implement physical plan for it, we get this error naturally.

Example:

```SQL
cd datafusion-cli 
cargo run
DataFusion CLI v15.0.0
❯ PREPARE my_plan(INT) AS SELECT $1
;
Internal("Unsupported logical plan: Prepare")
❯ 
```

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/prepare.slt:
```diff
@@ -0,0 +1,40 @@
+# PREPARE my_plan(STRING, STRING) AS SELECT * FROM (VALUES(1, $1), (2, $2)) AS t (num, letter);
+
+# And then we may want to add test_prepare_statement* here
+
```
Since we do not store the prepare logical plan anywhere and still throw error when we try to generate physical plan for it, this .stl tests are not that useful yet but they will be valuable after those are done

alamb replied:
Perhaps I can add a subtask to implement it under https://github.com/apache/arrow-datafusion/issues/4539 -- then in DataFusion we normally leave a link to the ticket as a comment

NGA-TRAN commented on datafusion/core/tests/sqllogictests/test_files/prepare.slt:
```diff
@@ -0,0 +1,40 @@
+1 jane smith 20 MA 100000.45 2000-11-12T00:00:00.000000000 99
+
+# TODO: support error instead of panicking
+# thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SQL(ParserError("Expected AS, found: SELECT"))', datafusion/core/tests/sqllogictests/src/main.rs:197:42
```
I hope someone will pick this todo in the future

alamb replied:
I think I fixed this in https://github.com/apache/arrow-datafusion/pull/4530 -- so I think if you merge this PR up to master again it will work

NGA-TRAN replied:
Oh yes, both ok and error work now. I will add more tests

NGA-TRAN replied:
Done

alamb commented on datafusion/expr/src/expr.rs:
```diff
@@ -244,6 +244,14 @@ pub enum Expr {
+    /// (e.g. `$foo` or `$1`)
+    Placeholder {
+        /// The identifier of the parameter (e.g, $1 or $foo)
+        id: String,
```
👍

alamb commented on datafusion/expr/src/logical_plan/builder.rs:
```diff
@@ -279,6 +281,14 @@ impl LogicalPlanBuilder {
         )?)))
     }
 
+    pub fn prepare(&self, name: String, data_types: Vec<DataType>) -> Result<Self> {
```
Perhaps we can add a docstring to this method

NGA-TRAN replied:
Done

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -16,7 +16,7 @@
 
 //! SQL Query Planner (produces logical plan from SQL AST)
-
+use log::debug;
```
👍  I have wanted some debug logging in the planner as well

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -1857,6 +1909,41 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+        param: String,
+        param_data_types: &[DataType],
+    ) -> Result<Expr> {
+        // Parse the placeholder as a number becasue it is the only support from sqlparser and postgres
```
```suggestion
        // Parse the placeholder as a number because it is the only support from sqlparser and postgres
```

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -1857,6 +1909,41 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+            Ok(index) => index - 1,
+            Err(_) => {
+                return Err(DataFusionError::Internal(format!(
+                    "Invalid placeholder: {}",
```
```suggestion
                    "Invalid placeholder, not a number: {}",
```

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -6144,6 +6247,185 @@ mod tests {
+        prepare_stmt_quick_test(sql, expected_plan, expected_dt);
+    }
+
+    #[test]
```
👍  very nice tests

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -6144,6 +6247,185 @@ mod tests {
+
+        let expected_plan = "Prepare: \"my_plan\" [Int32, Float64] \
+        \n  Projection: Int64(1) + $1 + $2\n    EmptyRelation";
+        let expected_dt = "[Int32, Float64]";
```
👍

alamb reviewed (approved):
Looks great to me -- thanks @NGA-TRAN 

I plan to wait for @mingmwang 's review prior to merging this PR but I think it could go in now

mingmwang commented on datafusion/expr/src/logical_plan/plan.rs:
```diff
@@ -1360,6 +1370,18 @@ pub struct CreateExternalTable {
+    /// The name of the statement
+    pub name: String,
+    /// Data types of the parameters ([`Expr::Placeholder`])
+    pub data_types: Vec<DataType>,
+    /// The logical plan of the statements
+    pub input: Arc<LogicalPlan>,
+}
+
```
I assume the data types Vec size is the same with the place holders in the input plan, but is there any check for this?

NGA-TRAN replied:
The data types of the placeholders are from the data types of the Vec here so they match. We do check if the Vec contains enough params, too. However, there are flexibility:
1. The length of the Vec can be longer than the number of the placeholders which will be fine and we have tests for this.
2. The data type of the Vec can be anything and we will use them for the placeholders, we do not check if the data types are compatible with the variables in the expression because: (i) we allow data type casting, and (2) when we reach the placeholders, we no longer have the context which variable/column/expression the place holder is used for. We do not want to add more context to backtrack which will cost compile time as well as complicated implementation.

However, I am working on https://github.com/apache/arrow-datafusion/issues/4550 that convert Prepare Logical Plan to a logical plan with all placeholders replaced with actual values. There, I will throw error if the data types provided do not work. We follow the same behavior of Postgres

mingmwang commented on datafusion/expr/src/utils.rs:
```diff
@@ -579,6 +580,13 @@ pub fn from_plan(
+        LogicalPlan::Prepare(Prepare {
+            name, data_types, ..
+        }) => Ok(LogicalPlan::Prepare(Prepare {
+            name: name.clone(),
+            data_types: data_types.clone(),
+            input: Arc::new(inputs[0].clone()),
+        })),
```
Is it possible and allowed here that the method passed in a totally different input plan ?

NGA-TRAN replied:
With the code we implement, the answer is no unless there are bugs

mingmwang commented:
BTW, do we support this case ?

Plain SQL:

`SELECT * FROM tab WHERE col_a is null;`

Prepared Stmt:

```
REPARE my_plan_1(INT) AS 
   SELECT * FROM tab WHERE col_a is $1;

```

NGA-TRAN commented on datafusion/sql/src/planner.rs:
```diff
@@ -1818,6 +1867,9 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                             Ok(Expr::Literal(ScalarValue::Null))
                         }
                         SQLExpr::Value(Value::Boolean(n)) => Ok(lit(n)),
+                        SQLExpr::Value(Value::Placeholder(param)) => {
+                            Self::create_placeholder_expr(param, param_data_types)
+                        }
```
I do not get the question but I think my answer for this https://github.com/apache/arrow-datafusion/pull/4490/files#r1043122807 probably answer this question, too

NGA-TRAN commented on datafusion/sql/src/planner.rs:
```diff
@@ -1857,6 +1909,41 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+        param: String,
+        param_data_types: &[DataType],
+    ) -> Result<Expr> {
+        // Parse the placeholder as a number because it is the only support from sqlparser and postgres
```
Done

NGA-TRAN reviewed (commented):
@alamb The PR is ready to master merge. I will resolve conflicts shortly but more conflicts may come if we do not merge soon 😄

alamb commented:
I plan to merge this as soon as the CI checks complete
